src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -1237,144 +1237,200 @@ Pipe MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
-    key_columns.reserve(sort_description.size());
+                case MergeTreeData::MergingParams::VersionedCollapsing:
+                    return std::make_shared<VersionedCollapsingTransform>(
+                        header, pipe.numOutputPorts(), sort_description, data.merging_params.sign_column, max_block_size);
 
-    for (auto & desc : sort_description)
-    {
-        if (!desc.column_name.empty())
-            key_columns.push_back(header.getPositionByName(desc.column_name));
-        else
-            key_columns.emplace_back(desc.column_number);
-    }
+                case MergeTreeData::MergingParams::Graphite:
+                    throw Exception("GraphiteMergeTree doesn't support FINAL", ErrorCodes::LOGICAL_ERROR);
+            }
 
-    pipe.addSimpleTransform([&](const Block & stream_header)
-    {
-        return std::make_shared<AddingSelectorTransform>(stream_header, num_streams, key_columns);
-    });
+            __builtin_unreachable();
+        };
 
-    pipe.transform([&](OutputPortRawPtrs ports)
-    {
-        Processors processors;
-        std::vector<OutputPorts::iterator> output_ports;
-        processors.reserve(ports.size() + num_streams);
-        output_ports.reserve(ports.size());
+        if (num_streams <= 1 || sort_description.empty())
+        {
+            pipe.addTransform(get_merging_processor());
+            partition_pipes.emplace_back(std::move(pipe));
+            continue;
+        }
+
+        ColumnNumbers key_columns;
+        key_columns.reserve(sort_description.size());
 
-        for (auto & port : ports)
+        for (auto & desc : sort_description)
         {
-            auto copier = std::make_shared<CopyTransform>(header, num_streams);
-            connect(*port, copier->getInputPort());
-            output_ports.emplace_back(copier->getOutputs().begin());
-            processors.emplace_back(std::move(copier));
+            if (!desc.column_name.empty())
+                key_columns.push_back(header.getPositionByName(desc.column_name));
+            else
+                key_columns.emplace_back(desc.column_number);
         }
 
-        for (size_t i = 0; i < num_streams; ++i)
+        pipe.addSimpleTransform([&](const Block & stream_header)
         {
-            auto merge = get_merging_processor();
-            merge->setSelectorPosition(i);
-            auto input = merge->getInputs().begin();
+            return std::make_shared<AddingSelectorTransform>(stream_header, num_streams, key_columns);
+        });
 
-            /// Connect i-th merge with i-th input port of every copier.
-            for (size_t j = 0; j < ports.size(); ++j)
+        pipe.transform([&](OutputPortRawPtrs ports)
+        {
+            Processors processors;
+            std::vector<OutputPorts::iterator> output_ports;
+            processors.reserve(ports.size() + num_streams);
+            output_ports.reserve(ports.size());
+
+            LOG_DEBUG(log, "Output ports size: {}", ports.size());
```

Comment:
Did you forget to remove debug output? If you want to leave it, please, add more context to message.